File: results/0.0/ZIP/zip-gpt-4o.c
#include <stdio.h>
#include <stdlib.h>
#include <hammer/hammer.h>

HParser *create_zip_parser() {
    HParser *signature = h_bits(32, false);
    HParser *version = h_bits(16, false);
    HParser *bit_flag = h_bits(16, false);
    HParser *compression_method = h_bits(16, false);
    HParser *mod_time = h_bits(16, false);
    HParser *mod_date = h_bits(16, false);
    HParser *crc32 = h_bits(32, false);
    HParser *compressed_size = h_bits(32, false);
    HParser *uncompressed_size = h_bits(32, false);
    HParser *file_name_length = h_bits(16, false);
    HParser *extra_field_length = h_bits(16, false);
    HParser *file_name = h_length_value(file_name_length, h_uint8());
    HParser *extra_field = h_length_value(extra_field_length, h_uint8());

    HParser *local_file_header = h_sequence(
        h_token((const uint8_t *)"\x50\x4b\x03\x04", 4),
        version,
        bit_flag,
        compression_method,
        mod_time,
        mod_date,
        crc32,
        compressed_size,
        uncompressed_size,
        file_name_length,
        extra_field_length,
        file_name,
        extra_field,
        NULL
    );

    HParser *central_directory_file_header = h_sequence(
        h_token((const uint8_t *)"\x50\x4b\x01\x02", 4),
        version,
        version,
        bit_flag,
        compression_method,
        mod_time,
        mod_date,
        crc32,
        compressed_size,
        uncompressed_size,
        file_name_length,
        extra_field_length,
        h_bits(16, false), // file comment length
        h_bits(16, false), // disk number start
        h_bits(16, false), // internal file attributes
        h_bits(32, false), // external file attributes
        h_bits(32, false), // relative offset of local header
        file_name,
        extra_field,
        h_length_value(h_bits(16, false), h_uint8()), // file comment
        NULL
    );

    HParser *end_of_central_directory_record = h_sequence(
        h_token((const uint8_t *)"\x50\x4b\x05\x06", 4),
        h_bits(16, false), // number of this disk
        h_bits(16, false), // disk where central directory starts
        h_bits(16, false), // number of central directory records on this disk
        h_bits(16, false), // total number of central directory records
        h_bits(32, false), // size of central directory
        h_bits(32, false), // offset of start of central directory
        h_length_value(h_bits(16, false), h_uint8()), // zip file comment
        NULL
    );

    HParser *zip_parser = h_many1(h_choice(
        local_file_header,
        central_directory_file_header,
        end_of_central_directory_record,
        NULL
    ));

    return zip_parser;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <zipfile>\n", argv[0]);
        return EXIT_FAILURE;
    }

    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        perror("Failed to open file");
        return EXIT_FAILURE;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    unsigned char *data = malloc(file_size);
    if (!data) {
        perror("Failed to allocate memory");
        fclose(file);
        return EXIT_FAILURE;
    }

    fread(data, 1, file_size, file);
    fclose(file);

    HParser *zip_parser = create_zip_parser();
    HParseResult *result = h_parse(zip_parser, data, file_size);

    if (result) {
        printf("ZIP file parsed successfully.\n");
        h_parse_result_free(result);
    } else {
        printf("Failed to parse ZIP file.\n");
    }

    free(data);
    h_parser_free(zip_parser);

    return EXIT_SUCCESS;
}

File: results/0.0/ZIP/zip-gpt-4o.rs
use nom::{
    bytes::complete::{tag, take},
    combinator::{map, map_opt, map_res},
    multi::many0,
    number::complete::{le_u16, le_u32, le_u8},
    sequence::tuple,
    IResult,
};
use std::fs::File;
use std::io::Read;
use std::path::Path;
use std::env;

#[derive(Debug)]
struct LocalFileHeader {
    version: u16,
    flags: u16,
    compression: u16,
    mod_time: u16,
    mod_date: u16,
    crc32: u32,
    compressed_size: u32,
    uncompressed_size: u32,
    file_name_length: u16,
    extra_field_length: u16,
    file_name: Vec<u8>,
    extra_field: Vec<u8>,
}

#[derive(Debug)]
struct CentralDirectoryFileHeader {
    version_made_by: u16,
    version_needed: u16,
    flags: u16,
    compression: u16,
    mod_time: u16,
    mod_date: u16,
    crc32: u32,
    compressed_size: u32,
    uncompressed_size: u32,
    file_name_length: u16,
    extra_field_length: u16,
    file_comment_length: u16,
    disk_number_start: u16,
    internal_file_attributes: u16,
    external_file_attributes: u32,
    relative_offset: u32,
    file_name: Vec<u8>,
    extra_field: Vec<u8>,
    file_comment: Vec<u8>,
}

#[derive(Debug)]
struct EndOfCentralDirectoryRecord {
    disk_number: u16,
    central_directory_disk_number: u16,
    central_directory_records_on_this_disk: u16,
    total_central_directory_records: u16,
    central_directory_size: u32,
    central_directory_offset: u32,
    comment_length: u16,
    comment: Vec<u8>,
}

fn parse_local_file_header(input: &[u8]) -> IResult<&[u8], LocalFileHeader> {
    let (input, _) = tag([0x50, 0x4b, 0x03, 0x04])(input)?;
    let (input, (version, flags, compression, mod_time, mod_date, crc32, compressed_size, uncompressed_size, file_name_length, extra_field_length)) =
        tuple((le_u16, le_u16, le_u16, le_u16, le_u16, le_u32, le_u32, le_u32, le_u16, le_u16))(input)?;
    let (input, file_name) = take(file_name_length)(input)?;
    let (input, extra_field) = take(extra_field_length)(input)?;

    Ok((
        input,
        LocalFileHeader {
            version,
            flags,
            compression,
            mod_time,
            mod_date,
            crc32,
            compressed_size,
            uncompressed_size,
            file_name_length,
            extra_field_length,
            file_name: file_name.to_vec(),
            extra_field: extra_field.to_vec(),
        },
    ))
}

fn parse_central_directory_file_header(input: &[u8]) -> IResult<&[u8], CentralDirectoryFileHeader> {
    let (input, _) = tag([0x50, 0x4b, 0x01, 0x02])(input)?;
    let (input, (version_made_by, version_needed, flags, compression, mod_time, mod_date, crc32, compressed_size, uncompressed_size, file_name_length, extra_field_length, file_comment_length, disk_number_start, internal_file_attributes, external_file_attributes, relative_offset)) =
        tuple((le_u16, le_u16, le_u16, le_u16, le_u16, le_u16, le_u32, le_u32, le_u32, le_u16, le_u16, le_u16, le_u16, le_u16, le_u32, le_u32))(input)?;
    let (input, file_name) = take(file_name_length)(input)?;
    let (input, extra_field) = take(extra_field_length)(input)?;
    let (input, file_comment) = take(file_comment_length)(input)?;

    Ok((
        input,
        CentralDirectoryFileHeader {
            version_made_by,
            version_needed,
            flags,
            compression,
            mod_time,
            mod_date,
            crc32,
            compressed_size,
            uncompressed_size,
            file_name_length,
            extra_field_length,
            file_comment_length,
            disk_number_start,
            internal_file_attributes,
            external_file_attributes,
            relative_offset,
            file_name: file_name.to_vec(),
            extra_field: extra_field.to_vec(),
            file_comment: file_comment.to_vec(),
        },
    ))
}

fn parse_end_of_central_directory_record(input: &[u8]) -> IResult<&[u8], EndOfCentralDirectoryRecord> {
    let (input, _) = tag([0x50, 0x4b, 0x05, 0x06])(input)?;
    let (input, (disk_number, central_directory_disk_number, central_directory_records_on_this_disk, total_central_directory_records, central_directory_size, central_directory_offset, comment_length)) =
        tuple((le_u16, le_u16, le_u16, le_u16, le_u32, le_u32, le_u16))(input)?;
    let (input, comment) = take(comment_length)(input)?;

    Ok((
        input,
        EndOfCentralDirectoryRecord {
            disk_number,
            central_directory_disk_number,
            central_directory_records_on_this_disk,
            total_central_directory_records,
            central_directory_size,
            central_directory_offset,
            comment_length,
            comment: comment.to_vec(),
        },
    ))
}

fn parse_zip(input: &[u8]) -> IResult<&[u8], (Vec<LocalFileHeader>, Vec<CentralDirectoryFileHeader>, EndOfCentralDirectoryRecord)> {
    let (input, local_file_headers) = many0(parse_local_file_header)(input)?;
    let (input, central_directory_file_headers) = many0(parse_central_directory_file_header)(input)?;
    let (input, end_of_central_directory_record) = parse_end_of_central_directory_record(input)?;

    Ok((input, (local_file_headers, central_directory_file_headers, end_of_central_directory_record)))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <zip-file>", args[0]);
        return;
    }

    let path = Path::new(&args[1]);
    let mut file = File::open(&path).expect("Failed to open file");
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).expect("Failed to read file");

    match parse_zip(&buffer) {
        Ok((_, (local_file_headers, central_directory_file_headers, end_of_central_directory_record))) => {
            println!("Local File Headers: {:?}", local_file_headers);
            println!("Central Directory File Headers: {:?}", central_directory_file_headers);
            println!("End of Central Directory Record: {:?}", end_of_central_directory_record);
        }
        Err(e) => eprintln!("Failed to parse ZIP file: {:?}", e),
    }
}